instrumentation/messaging/src/main/java/brave/messaging/MessagingAdapter.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2013-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.messaging;
+
+import brave.Span;
+import brave.internal.Nullable;
+
+/**
+ * @param <Chan> the type of the channel
+ * @param <Msg> the type of the message
+ * @param <C> the type that carriers the trace context, usually headers
+ */
+// abstract class instead of interface to allow method adds before Java 1.8
+public abstract class MessagingAdapter<Chan, Msg, C> {
+  // TODO: make some of these methods not abstract as they don't have meaning for all impls
```

Comment:
The only one that fit this description is `correlationId`. JMS and AMQP use it, but Kafka and RocketMQ don't. RocketMQ has a concept of `transactionId` that might replace it, but then I'd say each impl can add it's own additional identifiers/offsets.

Reply:
when you have thoughts like this. I would literally edit the comment and add to it until in the wiki or both. this helps others who miss a comment see the thought in progress. as soon as code drifts github doesn't show these threads anymore. make sense?

Reply:
of course. I'll do this.